common/cuda_hip/preconditioner/batch_isai_kernels.hpp.inc

```diff
@@ -0,0 +1,378 @@
+        }
+
+        subwarpgrp.sync();
+        const ValueType bot = subwarpgrp.shfl(sol, dense_col_idx);
+
+        if (local_id < dense_col_idx) {
+            sol = sol - bot * ele;
+        }
+    }
+
+    return sol;
+}
+
+
+template <typename Group, typename ValueType>
+__device__ __forceinline__ int choose_pivot_row(
+    Group subwarpgrp, const int diag_pos, const int size,
+    const ValueType* const local_row)
+{
+    const int local_id = subwarpgrp.thread_rank();
+    ValueType my_ele = local_row[diag_pos];
+    int my_idx = local_id;
+    const ValueType my_ele_1 = subwarpgrp.shfl(local_row[diag_pos], diag_pos);
+
+    if (local_id < diag_pos || local_id >= size) {
+        my_ele = my_ele_1;
+        my_idx = diag_pos;
+    }
+
+    subwarpgrp.sync();
+
+    for (int offset = subwarpgrp.size() / 2; offset > 0; offset /= 2) {
+        ValueType other_ele = subwarpgrp.shfl_down(my_ele, offset);
+        int other_idx = subwarpgrp.shfl_down(my_idx, offset);
+        if (abs(other_ele) > abs(my_ele)) {
+            my_ele = other_ele;
+            my_idx = other_idx;
+        }
+        subwarpgrp.sync();
+    }
+
+    // thread 0 would have the correct piv_row_idx and piv_ele
+    const int piv_row_idx = subwarpgrp.shfl(my_idx, 0);
+    return piv_row_idx;
+}
+
+template <typename Group, typename ValueType>
+__device__ __forceinline__ void swap_rows_and_rhs(
+    Group subwarpgrp, const int diag_pos, const int piv_row_idx, const int size,
+    ValueType* const local_row, ValueType& rhs)
+{
+    const int local_id = subwarpgrp.thread_rank();
+
+    for (int col = 0; col < size; col++) {
+        ValueType diag_tid_col_val = subwarpgrp.shfl(local_row[col], diag_pos);
+        ValueType piv_row_tid_col_val =
+            subwarpgrp.shfl(local_row[col], piv_row_idx);
+
+        if (local_id == diag_pos) {
+            local_row[col] = piv_row_tid_col_val;
+        }
+
+        if (local_id == piv_row_idx) {
+            local_row[col] = diag_tid_col_val;
+        }
+    }
+
+    ValueType diag_tid_rhs = subwarpgrp.shfl(rhs, diag_pos);
+    ValueType piv_row_tid_rhs = subwarpgrp.shfl(rhs, piv_row_idx);
+
+    if (local_id == diag_pos) {
+        rhs = piv_row_tid_rhs;
+    }
+
+    if (local_id == piv_row_idx) {
+        rhs = diag_tid_rhs;
+    }
+}
+
+
+template <typename Group, typename ValueType>
+__device__ __forceinline__ void row_transformation(Group subwarpgrp,
+                                                   const int diag_pos,
+                                                   const int size,
+                                                   ValueType* const local_row,
+                                                   ValueType& rhs)
+{
+    const int local_id = subwarpgrp.thread_rank();
+    const ValueType diag_ele = subwarpgrp.shfl(local_row[diag_pos], diag_pos);
+    // assert(diag_ele != zero<ValueType>());
+    assert(abs(diag_ele) !=
+           abs(zero<ValueType>()));  // temporary work-around for rocm-4.0
```

Comment:
Which is the least rocm version this works on ? I think we have already dropped 4.0 and only have 4.5 and upwards now.